contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,506 @@
+            self.check_role(caller, Role::Admin(this))?;
+
+            let pair = SwapPair::new(from, to);
+            self.swap_pairs.remove(&pair);
+
+            Self::emit_event(self.env(), Event::SwapPairRemoved(SwapPairRemoved { pair }));
+
+            Ok(())
+        }
+
+        /// Terminates the contract.
+        ///
+        /// Can only be called by the contract's Owner.
+        #[ink(message)]
+        pub fn terminate(&mut self) -> Result<(), DexError> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Owner(this))?;
+            self.env().terminate_contract(caller)
+        }
+
+        /// Returns own code hash
+        #[ink(message)]
+        pub fn code_hash(&self) -> Result<Hash, DexError> {
+            self.env()
+                .own_code_hash()
+                .map_err(|why| DexError::InkEnv(format!("Can't retrieve own code hash: {:?}", why)))
+        }
+
+        fn new_init(&mut self) {
+            self.access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+            self.swap_fee_percentage = 0;
+        }
+
+        /// Transfers a given amount of a PSP22 token to a specified using the callers own balance
+        fn transfer_tx(
+            &self,
+            token: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_SELECTOR))
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Transfers a given amount of a PSP22 token on behalf of a specified account to another account
+        ///
+        /// Will revert if not enough allowance was given to the caller prior to executing this tx
+        fn transfer_from_tx(
+            &self,
+            token: AccountId,
+            from: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_FROM_SELECTOR))
+                        .push_arg(from)
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true)) // needed for checking allowance before the actual tx
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
```

Comment:
nit: This seems like a very often-used util, maybe we should extract it to our own "lib".